### PR TITLE
8293416: ZGC: Set mark bit with unconditional atomic ops

### PR DESCRIPTION
**Pull request by heiher**
**Summary**
Support to set ZGC mark bit with unconditional atomic ops.

**Motivation**
ZGC currently modify mark-bitmap by a conditional atomic operation (cmpxchg). This way is not optimal, which will retry the loop when cmpxchg fails.

**Description**
First, This patch-set add an new unconditional atomic operation: Atomic::fetch_and_or, which is implemented in different ways for different CPU architectures:

* Exclusive access: Non-nested loop

```
retry:
  ll old_val, addr
  or new_val, old_val, set_val
  sc new_val, addr
  beq retry
```

* Atomic access: One instruction

```
ldset old_val, set_val, addr
```

* Generic: Fallback to cmpxchg or use c++ __atomic_fetch_or

**Testing**
* jtreg tests
* benchmark tests

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293416](https://bugs.openjdk.org/browse/JDK-8293416): ZGC: Set mark bit with unconditional atomic ops

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10182/head:pull/10182` \
`$ git checkout pull/10182`

Update a local copy of the PR: \
`$ git checkout pull/10182` \
`$ git pull https://git.openjdk.org/jdk pull/10182/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10182`

View PR using the GUI difftool: \
`$ git pr show -t 10182`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10182.diff">https://git.openjdk.org/jdk/pull/10182.diff</a>

</details>

### CODE REVIEW

**Comment by kimbarrett** [comment]
The Atomic changes being proposed here are a subset of
https://bugs.openjdk.org/browse/JDK-8293117 Add atomic bitset functions
which is in-progress.

File: src/hotspot/share/gc/z/zBitMap.inline.hpp [hunk 1]
```diff
@@ -47,26 +47,19 @@ inline bool ZBitMap::par_set_bit_pair_strong(idx_t bit, bool& inc_live) {
...
+    return false;
+  }
 
-    // The value changed, retry
-    old_val = cur_val;
-  } while (true);
+  const bm_word_t cur_val = Atomic::fetch_and_or(addr, pair_mask);
+  const bm_word_t marked_mask = bit_mask(bit);
+  inc_live = !(cur_val & marked_mask);
+  return inc_live;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zBitMap.inline.hpp:62]: I don't think this is correct -- previously, the return value and `inc_live` can take different values. For example, when the current thread successfully upgrades an obj from finalize-marked to strongly-marked, the return value is `true` while `inc_live == false`.

**Comment by heiher** [comment]
> The Atomic changes being proposed here are a subset of https://bugs.openjdk.org/browse/JDK-8293117 Add atomic bitset functions which is in-progress.

I'll depends on it.

Thank you.

File: src/hotspot/share/gc/z/zBitMap.inline.hpp (hunk 1 above)
Comment (heiher) [reply src/hotspot/share/gc/z/zBitMap.inline.hpp:62]: You're right. I misunderstood the paired mark. BitMap also seems to work with this approach as well. I'll do these in next version.

Thank you.

**Comment by heiher** [comment]
Hello,

I will rebase this patchset when JDK-8293117 (https://bugs.openjdk.org/browse/JDK-8293117)  is done. Atomic: Add bitset functions(https://github.com/openjdk/jdk/pull/10182/commits/dc15c68d1ff0e5c7445b333d45dc562881f5687e) contains a few architecture (not all) implementations, and only for test.

RFC,
Thanks~

**Comment by heiher** [comment]
## Micro Benchmark

Benchmark: https://gist.github.com/heiher/6137c8df7038af1a4186994894f2eb2b

### How to run

```bash
gcc -O3 -o bench atomic-flip-bit-bench.c -pthread
perf stat ./bench cas
perf stat ./bench amo
```
### Results

Real time elapsed in seconds (Less is better)

| CPU/Threads/Mode | Score 1 | Score 2 | Score 3 | Score 4 | Score 5 | Score 6 |
|--------|--------|--------|--------|--------|--------|--------|
| Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz |
| 8-Thread CAS | 6.20 | 5.26 | 4.80 | 5.94 | 6.22 | 6.09 |
| 8-Thread AMO | 1.20 | 1.30 | 1.13 | 1.33 | 1.32 | 1.17 |
| HUAWEI Kunpeng 920 @ 2.60GHz |
| 8-Thread CAS | 13.25 | 14.25 | 12.41 | 14.26 | 13.76 | 13.87 |
| 8-Thread AMO | 4.78 | 4.70 | 5.02 | 5.05 | 5.16 | 4.77 |
| 16-Thread CAS | 32.44 | 41.94 | 39.75 | 39.85 | 34.23 | 36.05 |
| 16-Thread AMO | 9.03 | 8.98 | 9.30 | 9.63 | 9.19 | 9.38 |

File: src/hotspot/share/gc/z/zBitMap.inline.hpp [hunk 2]
```diff
@@ -47,26 +47,29 @@ inline bool ZBitMap::par_set_bit_pair_strong(idx_t bit, bool& inc_live) {
...
+  }
+
+  // Slow path
+  cur_val = Atomic::fetch_and_or(addr, pair_mask);
+  new_val = cur_val | pair_mask;
+  if (new_val == cur_val) {
+    // Someone else beat us to it
+    inc_live = false;
+    return false;
+  }
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zBitMap.inline.hpp:67]: The fast/slow paths are almost identical. I believe some numbers would be nice to justify the presence of the fast path.

**Comment by heiher** [comment]
## Micro Benchmark 2

Benchmark: https://gist.github.com/heiher/a57c787e2f2d4f313df8cdf13086824e

### How to run

```bash
gcc -O3 -o bench atomic-mark-bench.c -pthread
perf stat ./bench norm
perf stat ./bench fast
```
### Results

Real time elapsed in seconds (Less is better)

| CPU/Threads/Mode | Score 1 | Score 2 | Score 3 | Score 4 | Score 5 | Score 6 |
|--------|--------|--------|--------|--------|--------|--------|
| Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz |
| 16-Thread NORM | 40.89 | 39.74 | 41.26 | 41.70 | 41.30 | 41.48 |
| 16-Thread FAST | 0.03 | 0.03 | 0.03 | 0.03 | 0.03 | 0.03 |
| HUAWEI Kunpeng 920 @ 2.60GHz |
| 16-Thread NORM | 98.44 | 94.73 | 99.05 | 93.95 | 97.34 | 94.95 |
| 16-Thread FAST | 0.11 | 0.11 | 0.11 | 0.11 | 0.11 | 0.11 |

File: src/hotspot/share/gc/z/zBitMap.inline.hpp (hunk 2 above)
Comment (heiher) [reply src/hotspot/share/gc/z/zBitMap.inline.hpp:67]: I think so. Atomic operations are more expensive than normal memory loads. This is why should split to fast and slow path, although they are almost Identical. The micro benchmark 2 shows it.

Comment (albertnetymk) [reply src/hotspot/share/gc/z/zBitMap.inline.hpp:67]: Thank you for the benchmark results; they are convincing, IMO.

**Pull request closed without merging**